Error computing dependencies on Python 3.7

In the release tests for v0.3 on Python 3.7, we found [a couple of test failures](https://github.com/diazona/setuptools-pyproject-migration/actions/runs/10543238199/job/29227734609#step:6:312) related to the dependencies of pytest.

<details>
<summary>Error log</summary>
```
=================================== FAILURES ===================================
_____________ TestExternalProject.test_dependencies[pytest==7.3.0] _____________

self = <test_distribution_packages.TestExternalProject object at 0x7fed85f79210>
expected = StandardMetadata(name='pytest', version=<Version('7.3.0')>, description='pytest: simple powerful testing with Python',..., 'Topic :: Utilities'], keywords=['test', 'unittest'], scripts={}, gui_scripts={}, dynamic=[], _metadata_version=None)
actual = StandardMetadata(name='pytest', version=<Version('7.3.0')>, description='pytest: simple powerful testing with Python',...'pytest': 'pytest:console_main', 'py.test': 'pytest:console_main'}, gui_scripts={}, dynamic=[], _metadata_version=None)

    def test_dependencies(self, expected: StandardMetadata, actual: StandardMetadata):
>       assert sorted(expected.dependencies, key=str) == sorted(actual.dependencies, key=str)
E       assert [<Requirement...=0.12')>, ...] == [<Requirement...2.0,>=0.12')>]
E         At index 0 diff: <Requirement('colorama; sys_platform == "win32"')> != <Requirement('iniconfig')>
E         Left contains 4 more items, first extra item: <Requirement('iniconfig')>
E         Full diff:
E           [
E         +  <Requirement('colorama; sys_platform == "win32"')>,
E         +  <Requirement('exceptiongroup>=1.0.0rc8; python_version < "3.11"')>,
E         +  <Requirement('importlib-metadata>=0.12; python_version < "3.8"')>,...
E         
E         ...Full output truncated (5 lines hidden), use '-vv' to show

/home/runner/work/setuptools-pyproject-migration/setuptools-pyproject-migration/tests/distribution/test_distribution_packages.py:174: AssertionError
________ TestExternalProject.test_optional_dependencies[pytest==7.3.0] _________

self = <test_distribution_packages.TestExternalProject object at 0x7fed85f79e10>
expected = StandardMetadata(name='pytest', version=<Version('7.3.0')>, description='pytest: simple powerful testing with Python',..., 'Topic :: Utilities'], keywords=['test', 'unittest'], scripts={}, gui_scripts={}, dynamic=[], _metadata_version=None)
actual = StandardMetadata(name='pytest', version=<Version('7.3.0')>, description='pytest: simple powerful testing with Python',...'pytest': 'pytest:console_main', 'py.test': 'pytest:console_main'}, gui_scripts={}, dynamic=[], _metadata_version=None)

    def test_optional_dependencies(self, expected: StandardMetadata, actual: StandardMetadata):
        # Check that the set of extras is the same
>       assert expected.optional_dependencies.keys() == actual.optional_dependencies.keys()
E       AssertionError: assert dict_keys(['testing']) == dict_keys(['testing', ''])
E         Full diff:
E         - dict_keys(['testing', ''])
E         ?                     ----
E         + dict_keys(['testing'])

/home/runner/work/setuptools-pyproject-migration/setuptools-pyproject-migration/tests/distribution/test_distribution_packages.py:178: AssertionError
```
</details>

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -48,8 +48,11 @@ jobs:
         - python: "3.13"
           platform: ubuntu-latest
           force-minimum-dependencies: false
-        # For testing forced minimum deps, use the latest stable version of Python
-        # on which those dependencies can be installed
+        # For testing forced minimum deps, use the newest and oldest stable
+        # versions of Python on which those dependencies can be installed.
+        - python: "3.8"
+          platform: ubuntu-latest
+          force-minimum-dependencies: true
         - python: "3.12"
           platform: ubuntu-latest
           force-minimum-dependencies: true

--- a/src/setuptools_pyproject_migration/__init__.py
+++ b/src/setuptools_pyproject_migration/__init__.py
@@ -1,5 +1,6 @@
 import configparser
 import itertools
+import logging
 import re
 import setuptools
 import sys
@@ -11,6 +12,9 @@ from setuptools_pyproject_migration._long_description import LongDescriptionMeta
 from setuptools_pyproject_migration._types import Contributor, Pyproject
 from tomlkit.api import Array, InlineTable
 from typing import Dict, List, Optional, Set, Tuple, TypeVar, Union
+
+
+_logger = logging.getLogger("setuptools_pyproject_migration")
 
 
 def _parse_entry_point(entry_point: str) -> Tuple[str, str]:
@@ -343,24 +347,39 @@ class WritePyproject(setuptools.Command):
             # version constraint string, which is exactly what we want here
             pyproject["project"]["requires-python"] = str(python_specifiers)
 
-        # NB: ensure a consistent alphabetical ordering of dependencies
-        dependencies = sorted(set(dist.install_requires))
-        if dependencies:
-            pyproject["project"]["dependencies"] = dependencies
+        _logger.debug("dist.install_requires: %r", dist.install_requires)
+        dependencies = set(dist.install_requires)
 
+        _logger.debug("dist.extras_require: %r", dist.extras_require)
         optional_dependencies: Dict[str, Set[str]] = {}
         for extra_dep_key, deps in dist.extras_require.items():
             extra, _, constraint = extra_dep_key.partition(":")
-            optional_dependencies.setdefault(extra, set())
+            _logger.debug("Handling extra=%r, constraint=%r", extra, constraint)
+            target: Set[str]
+            if extra:
+                _logger.debug("Adding to optional_dependencies[%r]", extra)
+                target = optional_dependencies.setdefault(extra, set())
+            else:
+                _logger.debug("Adding to dependencies")
+                target = dependencies
             for dep in deps:
                 if constraint:
-                    optional_dependencies[extra].add(f"{dep}; {constraint}")
+                    _logger.debug("Adding dependency %s with constraint %r", dep, constraint)
+                    target.add(f"{dep}; {constraint}")
                 else:
-                    optional_dependencies[extra].add(dep)
+                    _logger.debug("Adding dependency %s with no constraint", dep)
+                    target.add(dep)
+
+        if dependencies:
+            # NB: ensure a consistent alphabetical ordering of dependencies
+            sorted_dependencies = sorted(dependencies)
+            _logger.debug("Setting project.dependencies to: %r", sorted_dependencies)
+            pyproject["project"]["dependencies"] = sorted_dependencies
+
         if optional_dependencies:
-            pyproject["project"]["optional-dependencies"] = {
-                extra: sorted(deps) for extra, deps in optional_dependencies.items()
-            }
+            sorted_optional_dependencies = {extra: sorted(deps) for extra, deps in optional_dependencies.items()}
+            _logger.debug("Setting project.optional_dependencies to: %r", sorted_optional_dependencies)
+            pyproject["project"]["optional-dependencies"] = sorted_optional_dependencies
 
         entry_points = _generate_entry_points(dist.entry_points)
 

--- a/tests/distribution/test_distribution_packages.py
+++ b/tests/distribution/test_distribution_packages.py
@@ -24,7 +24,6 @@ import logging
 import packaging.markers
 import packaging.requirements
 import pytest
-import sys
 
 from typing import Iterator, List
 
@@ -51,8 +50,6 @@ def _setuptools_scm_version_conflict() -> bool:
     See `issue 145 <https://github.com/diazona/setuptools-pyproject-migration/issues/145>`_.
     """
 
-    if sys.version_info < (3, 12):
-        return False
     from test_support import importlib_metadata
     from packaging.version import Version
 

--- a/tests/test_optional_dependencies.py
+++ b/tests/test_optional_dependencies.py
@@ -205,3 +205,18 @@ def test_empty_extras(make_write_pyproject) -> None:
     cmd = make_write_pyproject(extras_require={})
     result = cmd._generate()
     assert "optional-dependencies" not in result["project"]
+
+
+def test_required_dependencies_with_constraints(make_write_pyproject) -> None:
+    """
+    Test that required dependencies with constraints are not added to the list
+    of optional dependencies.
+
+    This exercises a bug in setuptools<68.2 where required deps with constraints
+    get added to the list of optional dependencies with an empty extra.
+    """
+    dependencies: List[str] = ["holy-pin", 'holy-hand-grenade; python_version == "3.*"']
+    cmd = make_write_pyproject(install_requires=dependencies)
+    result = cmd._generate()
+    assert set(result["project"]["dependencies"]) == set(dependencies)
+    assert "optional-dependencies" not in result["project"]

